src/TierRegistry.sol

```diff
@@ -0,0 +1,330 @@
+        emit SetMaxDays(0, maxDays);
+
+        _grantRole(DEFAULT_ADMIN_ROLE, _initialRoleAdmin);
+        _grantRole(OWNER_ROLE, _initialOwner);
+        _grantRole(OPERATOR_ROLE, _initialOperator);
+    }
+
+    /*//////////////////////////////////////////////////////////////
+                               MODIFIERS
+    //////////////////////////////////////////////////////////////*/
+
+    modifier onlyOwner() {
+        if (!hasRole(OWNER_ROLE, msg.sender)) revert NotOwner();
+        _;
+    }
+
+    modifier onlyOperator() {
+        if (!hasRole(OPERATOR_ROLE, msg.sender)) revert NotOperator();
+        _;
+    }
+
+    /*//////////////////////////////////////////////////////////////
+                        STORAGE RENTAL LOGIC
+    //////////////////////////////////////////////////////////////*/
+
+    /**
+     * @inheritdoc ITierRegistry
+     */
+    function purchaseTier(uint256 fid, uint256 tier, uint256 forDays, address payer) external whenNotPaused {
+        uint256 pricePerDay = tokenPricePerDay[tier];
+        if (forDays == 0) revert InvalidAmount();
+        if (pricePerDay == 0) revert InvalidTier();
+        if (forDays < minDays) revert InvalidAmount();
+        if (forDays > maxDays) revert InvalidAmount();
+
+        uint256 cost = pricePerDay * forDays;
+
+        emit PurchasedTier(fid, tier, forDays);
+
+        IERC20 token = IERC20(paymentToken);
+        token.safeTransferFrom(payer, vault, cost);
+    }
+
+    /**
+     * @inheritdoc ITierRegistry
+     */
+    function batchPurchaseTiers(
+        uint256[] calldata fids,
+        uint256[] calldata tiers,
+        uint256[] calldata forDays,
+        address payer
+    ) external whenNotPaused {
+        if (fids.length == 0) revert InvalidBatchInput();
+        if (fids.length != tiers.length) revert InvalidBatchInput();
+        if (fids.length != forDays.length) revert InvalidBatchInput();
+
+        // Effects
+        uint256 totalCost;
+        for (uint256 i; i < fids.length; ++i) {
+            uint256 pricePerDay = tokenPricePerDay[tiers[i]];
+            uint256 numDays = forDays[i];
+            if (numDays == 0) revert InvalidAmount();
+            if (pricePerDay == 0) revert InvalidTier();
+            if (numDays < minDays) revert InvalidAmount();
+            if (numDays > maxDays) revert InvalidAmount();
+            totalCost += pricePerDay * numDays;
+        }
+
+        for (uint256 i; i < fids.length; ++i) {
+            emit PurchasedTier(fids[i], tiers[i], forDays[i]);
+        }
+
+        IERC20 token = IERC20(paymentToken);
+        token.safeTransferFrom(payer, vault, totalCost);
+    }
+
+    /*//////////////////////////////////////////////////////////////
+                         PERMISSIONED ACTIONS
+    //////////////////////////////////////////////////////////////*/
+
+    /**
+     * @inheritdoc ITierRegistry
+     */
+    function creditTier(uint256 fid, uint256 tier, uint256 forDays) external onlyOperator whenNotPaused {
```

Comment:
Not totally convinced we nee the crediting functionality. We do have it in the storage contract and it may make it easier to port over users who've already bought the pro tier but there is something nice from a transparency perspective about having a record of payment onchain for everybody who has the pro tier.